doc/neps/nep-0041-improved-dtype-support.rst

```diff
@@ -0,0 +1,393 @@
+to other updates in the recent past.
+However, if incompatibilities become more extensive then expected,
+a major NumPy release is an acceptable solution, although even then
+the vast majority of users shall not be affected.
+A transition requiring large code adaptation, similar to the Python 2 to 3
+transition is not anticipated and not covered by this NEP.
+
+Detailed below, we accept the following design considerations:
+
+1. Each basic datatype should be a class with most logic being implemented
+   as special methods on the class. In the C-API, these correspond to specific
+   slots.
+2. The current datatypes will be instances of these classes. Anything defined
+   on the instance, should instead be defined on the class (except things
+   instance information such as itemsize, byteorder, ...).
+4. All new API provided to the user will hide implementation as much as
+   possible. The public API should be identical but may be more limited,
+   to the API used to define the internal NumPy datatypes.
+5. The current NumPy scalars will *not* be instances of datatypes.
+3. The UFunc machinery will be changed to replace the current dispatching
+   and type resolution system. The old system should be (mostly) supported
+   as a legacy version for some time. This should thus not affect most users.
+   
+Additionally, as a general design principle, the addition of new user defined
+datatypes shall *not* change the behaviour of programs.
+For example ``common_dtype(a, b)`` must not be ``c`` unless ``a`` or ``b`` know
+that ``c`` exists.
+
+
+
+Detailed Description
+--------------------
+
+The following sections details some of the design decisions above and gives
+more details on potential user datatype use cases motivating the need for
+these changes.
+Since datatype changes touch a large part of code and behaviours, NEP 40
+reviews some of the concepts, issues, and the current implementation.
+
+
+Motivation and the Need for New User-Defined Datatypes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The current ecosystem has very few user defined datatypes using NumPy, the
+two most promient being: ``rational`` and ``quaternion``.
+These represent fairly simple datatypes which are not as strongly impacted
+but the current limitations.
+However, the number of currently available user defined datatypes
+is in strong contrast with the need for datatypes such as:
+
+* categorical types
+* bfloat16, used in deep learning
+* physical units (such as meters)
+* extending e.g. integer dtypes to have a sentinel NA value
+* geometrical objects [pygeos]_
+* datatypes for tracing/automatic differentiation
+* high or arbitrary precision math
+* ….
+
+Some of these are partially solved; for example unit capability is provided
+in ``astropy.units``, ``unyt``, or ``pint``. However, these have to subclass
+or wrap ``ndarray`` whereas special datatypes would provide a much more natural
+representation, and would immediately allow usage within tools such as
```

Comment:
This would be a great place to add some example code to show the more natural representation.

Reply:
Actually I think you may want two separate code examples: one aimed at users of a new dtype, and one aimed at authors of libraries that would want to create new dtypes.

Reply:
Well, I left those out intentionally. There are too many moving parts and things to bike shed, IMO.

What I want to convince you of here is:

1. We want "full featured dtypes"
2. We currently cannot have them
3. These are *necessary* decisions we can get out of the way before we do the detailed API design.

It may be that the text further down is too detailed in some cases, and thus hides that intention? Or do I need an example to show that `Dask` should simply work with a new `dtype`, but working with a new `array-like` such as `astropy.units` is probably less straight forward?

Reply:
Well, I tried to clarify the part of why I think subclasses are not a real solution a bit, maybe that helps. I am thinking of rather than adding a lot of new things, I may delete a bit of the discussions below (the API breakage is cleaned up a bit).